File: protocol-generator/dart/test/pre_admission_test.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:entity_core_protocol/entity_core_peer.dart';
import 'package:entity_core_protocol/src/codec/ecf.dart';
import 'package:entity_core_protocol/src/codec/ecf_value.dart';
import 'package:entity_core_protocol/src/errors.dart';
import 'package:entity_core_protocol/src/peer/wire.dart' as wire;
import 'package:test/test.dart';

/// §4.11 pre-admission refusals (0.8.2.25) — classification AND emission.
///
/// §4.11's rule has two halves and they fail differently.
///
///   * "The frame obligation belongs to the class" is WIRE-VISIBLE, and the two
///     non-conformant behaviours it names are distinct: DROPPING the frame (no
///     response, no close — "the weaker of the two precisely because nothing surfaces
///     it") and CLOSING WITH NO CODED FRAME (indistinguishable from a network fault,
///     and on a multiplexed connection it destroys unrelated ADMITTED requests). This
///     peer had BOTH before 0.8.2.25: the un-salvageable decode arm and the
///     non-EXECUTE root dropped, the oversize arm and the mid-frame EOF closed bare.
///   * "The CODE belongs to the cause [MUST]" is a MAPPING, and a mapping is exactly
///     the thing that regresses silently when a new failure joins an existing branch.
///
/// Both halves are covered: the mapping at the unit level, and the emission over a
/// real socket, because a green mapping over a transport that never calls it is the
/// `check_path_permission` shape all over again.
///
/// THE PINNED CHECK SET (778) HAS NO VECTOR ON THIS SURFACE.
///
/// EVERY SOCKET CASE CARRIES A POSITIVE CONTROL. A probe fails in the direction of
/// the answer it is looking for: a malformed frame that is malformed in a SECOND way
/// answers the code under measurement for the wrong reason.
///
/// THE READ DEADLINE IS AN ASSERTION, NOT A CONVENIENCE. §4.11's non-conformant
/// behaviour is NO RESPONSE, so a socket read with no deadline HANGS on the defect
/// instead of failing on it — and a hung test is strictly worse than a red one: it
/// reports nothing and blocks everything behind it. Every read below is under
/// `.timeout(...)`, and the per-test `timeout` keeps a wedged socket from taking the
/// suite with it.
void main() {
  const readDeadline = Duration(seconds: 5);
  final caseTimeout = Timeout(const Duration(seconds: 20));

  Uint8List framed(Uint8List payload) {
    final out = Uint8List(payload.length + 4);
    final n = payload.length;
    out[0] = (n >> 24) & 0xff;
    out[1] = (n >> 16) & 0xff;
    out[2] = (n >> 8) & 0xff;
    out[3] = n & 0xff;
    out.setRange(4, 4 + n, payload);
    return out;
  }

  Uint8List enc(EcfValue v) => Ecf.encodeOrThrow(v);

  /// A well-formed EXECUTE the peer MUST answer 200 — the positive control.
  Future<Uint8List> helloFrame() async {
    final ident = await Identity.ofSeed(Uint8List(32)..fillRange(0, 32, 0x2a));
    final hello = Entity.make(
        'system/protocol/connect/hello',
        cmap([
          'peer_id', ident.peerId,
          'nonce', cbytes(Uint8List(32)..fillRange(0, 32, 0x01)),
          'protocols', textArray(['entity-core/1.0']),
          'timestamp', EcfInt.of(1),
          'hash_formats', textArray(['ecfv1-sha256']),
          'key_types', textArray(['ed25519']),
        ]));
    final env = Envelope(wire.makeExecute(
        'ctl-1', 'system/protocol/connect', 'hello', hello));
    return framed(wire.frameOfEnvelope(env));
  }

  /// A frame whose ONLY defect is a CBOR tag inside an entity's `data` map.
  ///
  /// Hand-spliced, because this peer's encoder cannot emit a tag by construction —
  /// AND THE SPLICE IS ASSERTED. A mutation that is not verified to have landed is
  /// not a mutation: an unverified splice leaves the frame perfectly well-formed, the
  /// peer answers something correct to a question this test is not asking, and that
  /// reads as a peer that does not implement the branch.
  Uint8List taggedExecutePayload() {
    final root = Entity.make(
        'system/protocol/execute',
        cmap([
          'request_id', 'tag-1',
          'uri', 'system/tree',
          'operation', 'get',
          'params', wire.emptyParams().toCbor(),
          'extra', EcfInt.of(0),
        ]));
    final payload = enc(EcfMap.of({'root': root.toCbor()}));
    // text(5) "extra", then uint 0.
    final marker = Uint8List.fromList(
        [0x65, 0x65, 0x78, 0x74, 0x72, 0x61, 0x00]);
    int indexOf(Uint8List hay, Uint8List needle, [int from = 0]) {
      outer:
      for (var i = from; i + needle.length <= hay.length; i++) {
        for (var j = 0; j < needle.length; j++) {
          if (hay[i + j] != needle[j]) continue outer;
        }
        return i;
      }
      return -1;
    }

    final at = indexOf(payload, marker);
    expect(at, isNot(-1),
        reason: 'the splice target moved; the mutation is not a mutation');
    expect(indexOf(payload, marker, at + 1), equals(-1),
        reason: 'the splice target is ambiguous');
    final out = Uint8List(payload.length + 1)
      ..setRange(0, at + marker.length - 1, payload)
      ..[at + marker.length - 1] = 0xc1 // tag 1, in front of the uint 0
      ..[at + marker.length] = 0x00
      ..setRange(at + marker.length + 1, payload.length + 1, payload,
          at + marker.length);
    // The splice is asserted to BE a tag rejection before any socket sees it.
    expect(Ecf.decode(out).errorOrNull, isA<TagRejected>());
    return out;
  }

  Future<T> withPeer<T>(Future<T> Function(int port) body) async {
    final peer =
        await Peer.create(Uint8List(32)..fillRange(0, 32, 0x7b));
    final listener = await startListener(peer, 0);
    try {
      return await body(listener.port);
    } finally {
      await listener.close();
    }
  }

  /// Read `expect` framed responses off `s`, each under the deadline.
  Future<List<(int, String, String)>> readN(
      Stream<Uint8List> stream, int want) async {
    final acc = <int>[];
    final out = <(int, String, String)>[];
    await for (final chunk in stream.timeout(readDeadline,
        onTimeout: (sink) => sink.addError(StateError(
            'no response: the frame was DROPPED (section 4.11)')))) {
      acc.addAll(chunk);
      while (acc.length >= 4) {
        final n = (acc[0] << 24) | (acc[1] << 16) | (acc[2] << 8) | acc[3];
        if (acc.length < 4 + n) break;
        final env = wire.envelopeOfFrame(
            Uint8List.fromList(acc.sublist(4, 4 + n)));
        acc.removeRange(0, 4 + n);
        final res = wire.responseResult(env);
        out.add((
          wire.responseStatus(env),
          res?.text('code') ?? '',
          env.root.text('request_id') ?? ''
        ));
      }
      if (out.length >= want) break;
    }
    return out;
  }

  /// Send `frames` on ONE connection and read `want` responses.
  Future<List<(int, String, String)>> drive(
      List<Uint8List> frames, int want, int port) async {
    final s = await Socket.connect('127.0.0.1', port);
    final bc = s.asBroadcastStream();
    try {
      for (final f in frames) {
        s.add(f);
      }
      await s.flush();
      return await readN(bc, want);
    } finally {
      s.destroy();
    }
  }

  group('section 4.11 — the CODE belongs to the CAUSE (0.8.2.25, section 5.2a)', () {
    // Before 0.8.2.24 this peer answered `400 non_canonical_ecf` for every one of
    // these, which is the code-under-the-wrong-reason defect §5.2a names: a mis-keyed
    // `included` entry carries NO TAG, its encoding is canonical, and *re-encode* is
    // not the caller's remedy.
    final rows = <(Object, int, String)>[
      // §4.10(a), mood raised SHOULD -> MUST at 0.8.2.25 (N14).
      (const PayloadTooLarge('x'), 413, 'payload_too_large'),
      // §5.2a / §1.8 resolution integrity. 0.8.2.24 pins this and rules
      // non_canonical_ecf NON-CONFORMANT here.
      (const HashMismatch('x'), 400, 'hash_mismatch'),
      // ENTITY-CBOR-ENCODING §6.3 — the tag-policy arm keeps its own code.
      (const TagRejected('x'), 400, 'non_canonical_ecf'),
      // §4.7 / §4.11 framing arm: bytes that never become an Envelope.
      (const TruncatedInput('x'), 400, 'invalid_request'),
      // The row that makes the split load-bearing: a non-minimal head is
      // "non-canonical CBOR" BY NAME and is NOT the tag-policy arm.
      (const NonCanonicalEcf('x'), 400, 'invalid_request'),
      (const NonMinimalInt('x'), 400, 'invalid_request'),
      (const DuplicateKey('x'), 400, 'invalid_request'),
    ];

    test('maps each cause to its own status and code', () {
      expect(rows.length, equals(7)); // examined-N, not merely "no failures"
      for (final (cause, status, code) in rows) {
        // Wrapped AND bare: the decode path throws a DecodeRefusal carrying the
        // cause, while the transport's own arms pass the EntityError directly.
        for (final thrown in [cause, DecodeRefusal(cause as EntityError)]) {
          final r = wire.classifyPreAdmission(thrown);
          expect((r.status, r.code), equals((status, code)),
              reason: '${cause.runtimeType}');
          expect(r.message.codeUnits.every((c) => c < 128), isTrue,
              reason: 'a wire-visible message stays ASCII');
        }
      }
    });
  });

  group('section 4.11 — the decode boundary throws the CAUSE', () {
    final good = Entity.make('primitive/any', cmap(['x', 1]));

    test('a MIS-KEYED included entry throws HashMismatch', () {
      // The arc-probe B1/B2 input. The entry's ENCODING is canonical; what is false
      // is the claim the KEY makes.
      final payload = enc(EcfMap.of({
        'root': wire
            .makeExecute('t1', 'system/tree', 'get', wire.emptyParams())
            .toCbor(),
        'included': EcfMap([
          EcfEntry(cbytes(Uint8List(33)..fillRange(0, 33, 0x11)), good.toCbor())
        ]),
      }));
      Object? caught;
      try {
        wire.envelopeOfFrame(payload);
      } catch (e) {
        caught = e;
      }
      expect(caught, isA<DecodeRefusal>());
      expect((caught! as DecodeRefusal).cause, isA<HashMismatch>());
      final r = wire.classifyPreAdmission(caught);
      expect((r.status, r.code), equals((400, 'hash_mismatch')));
    });

    test('a non-binding carried content_hash throws HashMismatch', () {
      final m = EcfMap.of({
        'type': EcfText(good.type),
        'data': good.data(),
        'content_hash': cbytes(Uint8List(33)..fillRange(0, 33, 0x22)),
      });
      Object? caught;
      try {
        Entity.ofCbor(m);
      } catch (e) {
        caught = e;
      }
      expect(caught, isA<DecodeRefusal>());
      expect((caught! as DecodeRefusal).cause, isA<HashMismatch>());
    });

    test('a CBOR tag in a data-field position throws TagRejected', () {
      final payload = taggedExecutePayload();
      Object? caught;
      try {
        wire.envelopeOfFrame(payload);
      } catch (e) {
        caught = e;
      }
      expect(caught, isA<DecodeRefusal>());
      expect((caught! as DecodeRefusal).cause, isA<TagRejected>());
      final r = wire.classifyPreAdmission(caught);
      expect((r.status, r.code), equals((400, 'non_canonical_ecf')));
    });
  });

  group('section 4.11 — the frame obligation, over a real socket', () {
    test('the positive control answers 200 on its own', () async {
      // The control FIRST. If this ever fails, nothing below is a reading about the
      // peer — it is a reading about this file.
      await withPeer((port) async {
        expect(await drive([await helloFrame()], 1, port),
            equals([(200, '', 'ctl-1')]));
      });
    }, timeout: caseTimeout);

    test('a COMPLETE refused frame is answered, with the code of the cause, and the '
        'connection keeps serving', () async {
      final good = Entity.make('primitive/any', cmap(['x', 1]));
      final misKeyed = framed(enc(EcfMap.of({
        'root': wire
            .makeExecute('t1', 'system/tree', 'get', wire.emptyParams())
            .toCbor(),
        'included': EcfMap([
          EcfEntry(cbytes(Uint8List(33)..fillRange(0, 33, 0x11)), good.toCbor())
        ]),
      })));
      final tagged = framed(taggedExecutePayload());
      // A root that is neither EXECUTE nor EXECUTE_RESPONSE -> 400 invalid_request
      // (§3.3/§6.5 "Other type?", N12/N17). NOT a bare close, and NOT the silent drop
      // this peer used to answer it with.
      final otherRoot = framed(wire.frameOfEnvelope(Envelope(
          Entity.make('primitive/any', cmap(['request_id', 'x-1'])))));
      await withPeer((port) async {
        final got = await drive(
            [misKeyed, tagged, otherRoot, await helloFrame()], 4, port);
        expect(got[0], equals((400, 'hash_mismatch', 't1')));
        expect(got[1], equals((400, 'non_canonical_ecf', 'tag-1')));
        expect(got[2], equals((400, 'invalid_request', 'x-1')));
        expect(got[3], equals((200, '', 'ctl-1')),
            reason: 'still serving after three refusals');
      });
    }, timeout: caseTimeout);

    test('an UNCORRELATED refusal is a best-effort frame with an empty request_id',
        () async {
      // Where no `request_id` can be recovered, §4.11 prescribes "a best-effort coded
      // frame carrying no correlation" — an empty `request_id` IS that form. Guessing
      // one would correlate the refusal to somebody else's in-flight request.
      final notAnEnvelope = framed(enc(EcfMap.of({'nope': EcfInt.of(1)})));
      final garbage = framed(Uint8List.fromList([0xff, 0xff, 0xff, 0xff]));
      await withPeer((port) async {
        final got =
            await drive([notAnEnvelope, garbage, await helloFrame()], 3, port);
        expect(got[0], equals((400, 'invalid_request', '')));
        expect(got[1], equals((400, 'invalid_request', '')));
        expect(got[2], equals((200, '', 'ctl-1')));
      });
    }, timeout: caseTimeout);

    test('an OVERSIZE prefix is answered 413 BEFORE the close', () async {
      // §4.10(a) N14: SHOULD -> MUST. The over-size condition is detected at the
      // length prefix with the connection intact and nothing spent, so the 413 goes
      // out FIRST and the close comes after — the close is now IN ADDITION to the
      // frame, not instead of it. The peer's own listener is the control: it keeps
      // serving other connections.
      await withPeer((port) async {
        final s = await Socket.connect('127.0.0.1', port);
        final bc = s.asBroadcastStream();
        try {
          final big = wire.maxFrame + 1;
          s.add(Uint8List.fromList([
            (big >> 24) & 0xff,
            (big >> 16) & 0xff,
            (big >> 8) & 0xff,
            big & 0xff
          ])); // prefix only; no body ever sent
          await s.flush();
          final got = await readN(bc, 1);
          expect(got.length, equals(1));
          expect(got[0], equals((413, 'payload_too_large', '')),
              reason: 'no id was ever readable: the best-effort form');
        } finally {
          s.destroy();
        }
        expect(await drive([await helloFrame()], 1, port),
            equals([(200, '', 'ctl-1')]),
            reason: 'the listener survived the refusal');
      });
    }, timeout: caseTimeout);

    test('a TRUNCATED frame is answered', () async {
      // A declared body that never arrives — §4.11's framing arm names this input
      // outright. The write side is shut down so the peer sees EOF mid-frame rather
      // than an idle connection.
      //
      // THIS IS THE CASE THE MANAGED-RUNTIME WARNING IS ABOUT: the response can only
      // be written AFTER the client's FIN, so it fails on any runtime whose socket
      // layer closes the write side on a remote half-close. Node needed
      // `allowHalfOpen: true` and the BEAM needed `exit_on_close: false`; Dart's
      // `Socket` keeps its write side open until destroyed, which this case is what
      // MEASURES rather than assumes.
      await withPeer((port) async {
        final s = await Socket.connect('127.0.0.1', port);
        final bc = s.asBroadcastStream();
        try {
          s.add(Uint8List.fromList(
              [0x00, 0x00, 0x10, 0x00, 0xa1])); // declared 4096, sent 1
          await s.flush();
          await s.close(); // half-close: FIN, write side of OUR socket
          final got = await readN(bc, 1);
          expect(got.length, equals(1));
          expect(got[0], equals((400, 'invalid_request', '')));
        } finally {
          s.destroy();
        }
        expect(await drive([await helloFrame()], 1, port),
            equals([(200, '', 'ctl-1')]));
      });
    }, timeout: caseTimeout);

    test('a PARTIAL LENGTH PREFIX is answered', () async {
      // The arm the vanguard's first driver had no case for: its "truncated frame"
      // case sent a COMPLETE 4-byte prefix, so truncation was caught in the body read
      // and the prefix discrimination had nothing driving it.
      await withPeer((port) async {
        final s = await Socket.connect('127.0.0.1', port);
        final bc = s.asBroadcastStream();
        try {
          s.add(Uint8List.fromList([0x00, 0x00])); // two bytes of four
          await s.flush();
          await s.close();
          final got = await readN(bc, 1);
          expect(got.length, equals(1));
          expect(got[0], equals((400, 'invalid_request', '')));
        } finally {
          s.destroy();
        }
      });
    }, timeout: caseTimeout);

    test('a CLEAN close at a frame boundary is owed NOTHING', () async {
      // The differential for the two cases above, and the direction that would be
      // wrong in the expensive way: answering 400 to every peer that simply hangs up.
      // A complete request, its response read, then a clean half-close with an empty
      // reassembly window — the peer must send nothing further.
      await withPeer((port) async {
        final s = await Socket.connect('127.0.0.1', port);
        final bc = s.asBroadcastStream();
        final extra = <(int, String, String)>[];
        try {
          s.add(await helloFrame());
          await s.flush();
          expect(await readN(bc, 1), equals([(200, '', 'ctl-1')]));
          await s.close();
          // Anything arriving in the next second is a refusal that should not exist.
          final more = readN(bc, 1).then(extra.addAll).catchError((_) {});
          await Future.any(
              [more, Future<void>.delayed(const Duration(seconds: 1))]);
          expect(extra, isEmpty,
              reason: 'a clean EOF at a frame boundary is an ordinary close');
        } finally {
          s.destroy();
        }
      });
    }, timeout: caseTimeout);

    test('a pre-admission refusal is NEVER silence', () async {
      // The class obligation, stated once as its own assertion rather than inferred
      // from the rows above: EVERY pre-admission cause puts a frame on the wire.
      //
      // Dropping is §4.11's other non-conformant behaviour and is "the weaker of the
      // two precisely because nothing surfaces it" — a `continue` with no write looks
      // exactly like a peer that is merely slow, and the caller learns nothing until
      // its own §6.11(c) deadline. A deadline expiry here IS that failure.
      final causes = <Uint8List>[
        framed(enc(EcfMap.of({'nope': EcfInt.of(1)}))),
        framed(Uint8List.fromList([0xff, 0xff, 0xff, 0xff])),
        framed(Uint8List(0)),
        framed(wire.frameOfEnvelope(
            Envelope(Entity.make('primitive/any', emptyMap())))),
      ];
      expect(causes.length, equals(4));
      await withPeer((port) async {
        final got = await drive(causes, causes.length, port);
        expect(got.length, equals(4));
        for (final (status, code, _) in got) {
          expect(status, equals(400));
          expect(code, isNotEmpty,
              reason: 'every pre-admission refusal is coded, none is silence');
        }
      });
    }, timeout: caseTimeout);
  });
}

File: protocol-generator/typescript/test/pre-admission.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import net from "node:net";
import {
  Ecf,
  Entity,
  Envelope,
  ExecuteResponse,
  HashMismatchError,
  Peer,
  SeedPolicy,
  TagRejectedError,
} from "../src/index.js";
import { encode } from "../src/codec/canonical-cbor.js";
import { ecfBytes, ecfMap, ecfPreEncoded, ecfText } from "../src/codec/ecf-value.js";
import {
  FrameTooLargeError,
  framingRefusal,
  preAdmissionRefusal,
  TruncatedFrameError,
  writeFrame,
} from "../src/transport/frame-codec.js";

/**
 * §4.11 pre-admission refusals (0.8.2.25).
 *
 * §4.11's rule has two parts and they fail differently. *"A peer that refuses a frame
 * pre-admission MUST put a coded EXECUTE_RESPONSE on the wire [MUST]"* is the OBLIGATION,
 * wire-visible, and it is the half that was broken here — this peer had BOTH of the
 * behaviours §4.11 names separately as non-conformant: DROPPING the frame (a partial
 * trailing frame and an un-salvageable decode) and CLOSING with no coded frame (an
 * oversize prefix, and the non-EXECUTE root's bare `break`). *"The CODE belongs to the
 * cause [MUST]"* is a MAPPING, and a mapping is exactly the thing that regresses silently
 * when a new failure joins an existing branch.
 *
 * §4.9(c)'s deliver-or-signal rule is scoped to *"every request the peer ADMITS"* and
 * reaches none of these, which is why §4.11 exists. The pinned check set (778) has no
 * vector on this surface, so the coverage is authored here rather than inherited.
 */

const TIMEOUT = 10_000;

/** Start a peer on an ephemeral port and return it with a raw client socket. */
async function connectRaw(): Promise<{ peer: Peer; sock: net.Socket }> {
  const peer = new Peer({ seedPolicy: SeedPolicy.debugOpen() });
  const port = await peer.listen(0);
  const sock = await new Promise<net.Socket>((resolve, reject) => {
    const s = net.createConnection({ host: "127.0.0.1", port }, () => resolve(s));
    s.once("error", reject);
  });
  return { peer, sock };
}

/** Read the next complete frame off a raw socket and decode it as a coded refusal. */
function readRefusal(sock: net.Socket): Promise<{ status: number; code: string; requestId: string }> {
  return new Promise((resolve, reject) => {
    let buf = new Uint8Array(0);
    const timer = setTimeout(
      () => reject(new Error("no coded EXECUTE_RESPONSE arrived: the peer answered with silence or a bare close")),
      TIMEOUT,
    );
    const done = (v: { status: number; code: string; requestId: string }): void => {
      clearTimeout(timer);
      sock.removeListener("data", onData);
      resolve(v);
    };
    const onData = (chunk: Buffer): void => {
      const next = new Uint8Array(buf.length + chunk.length);
      next.set(buf);
      next.set(chunk, buf.length);
      buf = next;
      if (buf.length < 4) return;
      const len = new DataView(buf.buffer, buf.byteOffset, 4).getUint32(0, false);
      if (buf.length < 4 + len) return;
      try {
        const env = Envelope.decode(buf.slice(4, 4 + len));
        assert.equal(
          env.root.type,
          "system/protocol/execute/response",
          "§4.11 requires a coded EXECUTE_RESPONSE",
        );
        const response = new ExecuteResponse(env.root);
        done({
          status: response.statusCode,
          code: Ecf.optText(response.result.data, "code") ?? "",
          requestId: response.requestId,
        });
      } catch (e) {
        clearTimeout(timer);
        reject(e);
      }
    };
    sock.on("data", onData);
    sock.once("close", () => {
      if (buf.length === 0) {
        clearTimeout(timer);
        reject(new Error("the peer closed with no coded frame — §4.11's second non-conformant behaviour"));
      }
    });
  });
}

/** A well-formed frame whose `included` entry is filed under a hash that is not its own. */
function misKeyedFrame(requestId: string): Uint8Array {
  const good = Entity.create("primitive/any", Ecf.map(["x", Ecf.uint(1n)]));
  const root = Entity.create(
    "system/protocol/execute",
    Ecf.map(["request_id", Ecf.text(requestId)], ["uri", Ecf.text("system/tree")], ["operation", Ecf.text("get")]),
  );
  return encode(
    ecfMap([
      [ecfText("root"), ecfPreEncoded(root.wireBytes)],
      [ecfText("included"), ecfMap([[ecfBytes(new Uint8Array(33).fill(0x11)), ecfPreEncoded(good.wireBytes)]])],
    ]),
  );
}

// ── the CLASSIFICATION half ──────────────────────────────────────────────────

// §4.11's table, one row at a time. "A single code for the class would answer an honest
// caller under the wrong reason and send them to the wrong layer."
test("§4.11 — the refusal CODE is the cause's, not the class's", () => {
  // §4.10(a), mood raised to MUST at 0.8.2.25 (N14).
  assert.deepEqual(preAdmissionRefusal(new FrameTooLargeError("x")).status, 413);
  assert.equal(preAdmissionRefusal(new FrameTooLargeError("x")).code, "payload_too_large");
  // §5.2a / §1.8 resolution integrity. 0.8.2.24 pins this and rules non_canonical_ecf
  // NON-CONFORMANT here.
  assert.equal(preAdmissionRefusal(new HashMismatchError("x")).code, "hash_mismatch");
  assert.equal(preAdmissionRefusal(new HashMismatchError("x")).status, 400);
  // ENTITY-CBOR-ENCODING §5.4 — the tag-policy arm keeps its own code.
  assert.equal(preAdmissionRefusal(new TagRejectedError("x")).code, "non_canonical_ecf");
  // §4.7 / §4.11 framing arm: bytes that never become an Envelope.
  assert.equal(preAdmissionRefusal(new TruncatedFrameError("x")).code, "invalid_request");
  assert.equal(preAdmissionRefusal(new Error("anything else")).code, "invalid_request");
  // Every message is ASCII (RULE H: a wire-visible string is CBOR-text-encoded).
  for (const e of [new FrameTooLargeError("x"), new HashMismatchError("x"), new TagRejectedError("x"), new Error("x")]) {
    const { message } = preAdmissionRefusal(e);
    assert.ok([...message].every((c) => c.charCodeAt(0) < 128), `non-ASCII in a wire message: ${message}`);
  }

  // framingRefusal separates "owed a frame" from "the connection simply ended".
  assert.equal(framingRefusal(new FrameTooLargeError("x")), true);
  assert.equal(framingRefusal(new TruncatedFrameError("x")), true);
  assert.equal(framingRefusal(new Error("socket hung up")), false);
  assert.equal(framingRefusal(new HashMismatchError("x")), false);
});

// The two decode-boundary CAUSES must reach the classifier as DIFFERENT error types.
// Before 0.8.2.24 this peer answered `400 non_canonical_ecf` for every one of them, which
// is the code-under-the-wrong-reason defect §5.2a names: a mis-keyed `included` entry
// carries no tag, its encoding is canonical, and *re-encode* is not the caller's remedy.
test("§5.2a — the decode boundary distinguishes resolution integrity from structure", () => {
  assert.throws(
    () => Envelope.decode(misKeyedFrame("t1")),
    HashMismatchError,
    "a mis-keyed included entry is a RESOLUTION-INTEGRITY fault",
  );
  // A CORRECTLY keyed entry whose entity carries a wrong content_hash is the same class
  // (§1.8 item 1) and takes the same code.
  const tampered = encode(
    ecfMap([
      [ecfText("type"), ecfText("primitive/any")],
      [ecfText("data"), ecfPreEncoded(encode(Ecf.map(["x", Ecf.uint(1n)])))],
      // 0x00 = the ecfv1-sha256 format byte, then 32 wrong digest bytes. The format code
      // must be SUPPORTED or the decode fails one rung earlier on
      // `unsupported_content_hash_format`, which is a different cause and a different
      // code — the first cut of this fixture used 0xf2 throughout and measured that
      // instead, which is the "a probe fails in the direction of the answer it is looking
      // for" trap in a fixture.
      [ecfText("content_hash"), ecfBytes(Uint8Array.from([0x00, ...new Uint8Array(32).fill(0x22)]))],
    ]),
  );
  assert.throws(() => Entity.decode(tampered), HashMismatchError);

  // STRUCTURAL faults are NOT HashMismatchError. THIS IS THE DISCRIMINATOR: if both
  // causes collapsed into one type the assertions above would pass vacuously.
  const noType = encode(ecfMap([[ecfText("data"), ecfPreEncoded(encode(Ecf.uint(1n)))]]));
  assert.throws(() => Entity.decode(noType), (e: unknown) => !(e instanceof HashMismatchError));

  // And the WELL-FORMED envelope must still decode, or every case above is satisfied by a
  // decoder that refuses everything.
  const good = Entity.create("primitive/any", Ecf.map(["x", Ecf.uint(1n)]));
  const root = Entity.create("system/protocol/execute", Ecf.map(["request_id", Ecf.text("t1")]));
  const wellFormed = encode(
    ecfMap([
      [ecfText("root"), ecfPreEncoded(root.wireBytes)],
      [ecfText("included"), ecfMap([[ecfBytes(good.contentHash), ecfPreEncoded(good.wireBytes)]])],
    ]),
  );
  const env = Envelope.decode(wellFormed);
  assert.equal(env.find(good.contentHash) !== undefined, true, "the included entity did not survive decode");
});

// ── the OBLIGATION half, over a real socket ──────────────────────────────────

// An oversize length prefix — §4.10(a), mood raised to MUST at 0.8.2.25 (N14). The body is
// never drained, so the framing is lost and the peer closes afterwards; the close is now
// IN ADDITION TO the frame rather than instead of it.
test("§4.11 — an oversize length prefix is answered 413 before the close", async () => {
  const { peer, sock } = await connectRaw();
  try {
    // 32 MiB declared, nothing sent. The bound is read from the PREFIX, so no body needs
    // to exist for the refusal to fire — that is the whole point of checking it there.
    const prefix = new Uint8Array(4);
    new DataView(prefix.buffer).setUint32(0, 0x0200_0000, false);
    sock.write(prefix);
    const { status, code, requestId } = await readRefusal(sock);
    assert.equal(status, 413);
    assert.equal(code, "payload_too_large");
    // §4.11's best-effort UNCORRELATED form: there is no request_id to recover from a
    // frame whose body never arrived, and guessing one would correlate the refusal to
    // somebody else's in-flight request.
    assert.equal(requestId, "");
  } finally {
    sock.destroy();
    await peer.dispose();
  }
});

// A length prefix that declares more than is sent — §4.11's framing arm, "a length prefix
// that never completes" -> 400 invalid_request. The CONTROL is the next test: both end the
// stream, and only one is owed a frame.
test("§4.11 — a truncated frame is answered 400 invalid_request", async () => {
  const { peer, sock } = await connectRaw();
  try {
    sock.write(Buffer.from([0x00, 0x00, 0x10, 0x00, 0xa1]));
    sock.end();
    const { status, code } = await readRefusal(sock);
    assert.equal(status, 400);
    assert.equal(code, "invalid_request");
  } finally {
    sock.destroy();
    await peer.dispose();
  }
});

// THE CONTROL. A clean EOF at a frame boundary is an ordinary hangup, NOT a refusal, and
// is owed nothing. Without this row every assertion above is satisfied by a peer that
// answers 400 to everyone who hangs up, which would be a new defect rather than a fix.
test("§4.11 control — a clean close at a frame boundary is not a refusal", async () => {
  const { peer, sock } = await connectRaw();
  try {
    const bytes: number[] = [];
    sock.on("data", (c: Buffer) => bytes.push(...c));
    sock.end();
    await new Promise((r) => setTimeout(r, 400));
    assert.equal(bytes.length, 0, "a clean close was answered with bytes; it is not a refusal");
  } finally {
    sock.destroy();
    await peer.dispose();
  }
});

// A mis-keyed `included` entry — §5.2a (0.8.2.24 N4/N5) pins 400 hash_mismatch here and
// rules 400 non_canonical_ecf NON-CONFORMANT. The frame is COMPLETE, so the peer answers
// and KEEPS SERVING; the correlated id proves the salvage path ran.
test("§5.2a — a mis-keyed included entry is 400 hash_mismatch and the connection survives", async () => {
  const { peer, sock } = await connectRaw();
  try {
    const frame = misKeyedFrame("mk-1");
    await writeFrame(sock, frame);
    const first = await readRefusal(sock);
    assert.equal(first.status, 400);
    assert.equal(first.code, "hash_mismatch");
    assert.equal(first.requestId, "mk-1", "a complete frame's request_id is recoverable, so the refusal correlates");

    // AND THE CONNECTION KEEPS SERVING. A decode refusal on a complete frame does not
    // desynchronize the stream, so answering it and closing would be the cascade this
    // peer already paid 81 FAILs for. Send it again and expect a second answer.
    await writeFrame(sock, frame);
    const second = await readRefusal(sock);
    assert.equal(second.code, "hash_mismatch");
  } finally {
    sock.destroy();
    await peer.dispose();
  }
});

// §6.5's "Other type?" arm as rewritten at 0.8.2.25 (N12/N17): "400 invalid_request, coded
// frame; MAY then close. NOT a bare close." §3.3 previously read "the connection MUST be
// closed", assigning no code and requiring no frame, and this reader did exactly that — a
// bare `break`. §9.1's floor row that mandated it was REPLACED at the same revision (N18).
test("§4.11 — a non-EXECUTE root is answered 400 invalid_request, not closed on", async () => {
  const { peer, sock } = await connectRaw();
  try {
    const root = Entity.create("primitive/any", Ecf.map(["request_id", Ecf.text("x-1")]));
    await writeFrame(sock, new Envelope(root, []).encode());
    const { status, code, requestId } = await readRefusal(sock);
    assert.equal(status, 400);
    assert.equal(code, "invalid_request");
    // Correlated where the id is recoverable; the uncorrelated frame is the FALLBACK.
    assert.equal(requestId, "x-1");
  } finally {
    sock.destroy();
    await peer.dispose();
  }
});
